### ISSUE TITLE
Maven error with sub modules: parent pom not found

### ISSUE THREAD

**Issue by enridaga** [SPARQL-Anything/sparql.anything#420]
After a global build, I was trying to generate the CLI jar from the sub-module (without re-running everything) and stumbled into this issue.

Cannot build sub-module since when the `revision` variable was introduced. Strangely, maven doesn't use the variable within the local pom before searching for the parent pom. Logically, it could make sense because the parent pom is supposed to be loaded before the local one. However, it is quite handy having that variable and being able to make temporary releases without changes to each one of the pom.

This is what I get:

```bash
$ mvn install
[INFO] Scanning for projects...
[INFO] 
[INFO] -----------< io.github.sparql-anything:sparql-anything-cli >------------
[INFO] Building sparql-anything-cli 0.9.0-SNAPSHOT
[INFO] --------------------------------[ jar ]---------------------------------
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  0.319 s
[INFO] Finished at: 2023-11-06T09:20:06Z
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal on project sparql-anything-cli: Could not resolve dependencies for project io.github.sparql-anything:sparql-anything-cli:jar:0.9.0-SNAPSHOT: Failed to collect dependencies at io.github.sparql-anything:sparql-anything-model:jar:0.9.0-SNAPSHOT: Failed to read artifact descriptor for io.github.sparql-anything:sparql-anything-model:jar:0.9.0-SNAPSHOT: io.github.sparql-anything:sparql-anything-parent:pom:${revision} was not found in https://repo.maven.apache.org/maven2 during a previous attempt. This failure was cached in the local repository and resolution is not reattempted until the update interval of central has elapsed or updates are forced -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/DependencyResolutionException
```

It may also depend on my maven version:

```bash
Apache Maven 3.8.5 (3599d3414f046de2324203b78ddcf9b5e4388aa0)
Maven home: /usr/local/Cellar/maven/3.8.5/libexec
Java version: 14.0.2, vendor: Oracle Corporation, runtime: /Library/Java/JavaVirtualMachines/openjdk-14.0.2.jdk/Contents/Home
Default locale: en_GB, platform encoding: UTF-8
OS name: "mac os x", version: "10.16", arch: "x86_64", family: "mac"
```

Can anybody else try this?

**Comment by enridaga** [comment]
This also happens when specifying the revision:

```bash
$ mvn install -Drevision=0.9.0-SNAPSHOT
[INFO] Scanning for projects...
[INFO] 
[INFO] -----------< io.github.sparql-anything:sparql-anything-cli >------------
[INFO] Building sparql-anything-cli 0.9.0-SNAPSHOT
[INFO] --------------------------------[ jar ]---------------------------------
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  0.250 s
[INFO] Finished at: 2023-11-06T10:13:42Z
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal on project sparql-anything-cli: Could not resolve dependencies for project io.github.sparql-anything:sparql-anything-cli:jar:0.9.0-SNAPSHOT: Failed to collect dependencies at io.github.sparql-anything:sparql-anything-model:jar:0.9.0-SNAPSHOT: Failed to read artifact descriptor for io.github.sparql-anything:sparql-anything-model:jar:0.9.0-SNAPSHOT: io.github.sparql-anything:sparql-anything-parent:pom:${revision} was not found in https://repo.maven.apache.org/maven2 during a previous attempt. This failure was cached in the local repository and resolution is not reattempted until the update interval of central has elapsed or updates are forced -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/DependencyResolutionException
```

**Comment by luigi-asprino** [comment]
Hi @enridaga,

 `mvn clean install -Dgenerate-cli-jar=true` apparently works for me (for the whole project - not just submodule)

**Comment by enridaga** [comment]
> Hi @enridaga,
> 
> `mvn clean install -Dgenerate-cli-jar=true` apparently works for me (for the whole project - not just submodule)

For me as well, can you try going into `sparql-anything-cli` and run from there after you do a full build? The build go nicely from the reactor pom but fails when trying to build any of the submodules because failing to locate the parent pom, whose version is referenced by the variable.

**Comment by enridaga** [comment]
Also, no joy after upgrading but with a slighly different message:

```bash
$ mvn -version
Apache Maven 3.9.5 (57804ffe001d7215b5e7bcb531cf83df38f93546)
Maven home: /usr/local/Cellar/maven/3.9.5/libexec
Java version: 14.0.2, vendor: Oracle Corporation, runtime: /Library/Java/JavaVirtualMachines/openjdk-14.0.2.jdk/Contents/Home
Default locale: en_GB, platform encoding: UTF-8
OS name: "mac os x", version: "10.16", arch: "x86_64", family: "mac"
(base) MCL293239:sparql-anything-cli ed4565$ mvn -Drevision=0.9.0-SNAPSHOT install
[INFO] Scanning for projects...
[ERROR] [ERROR] Some problems were encountered while processing the POMs:
[FATAL] 'parent.version' is missing. @ line 22, column 13
 @ 
[ERROR] The build could not read 1 project -> [Help 1]
[ERROR]   
[ERROR]   The project io.github.sparql-anything:sparql-anything-cli:${revision} (/Users/ed4565/Development/sparql-anything/sparql.anything/sparql-anything-cli/pom.xml) has 1 error
[ERROR]     'parent.version' is missing. @ line 22, column 13
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/ProjectBuildingException

```

**Comment by luigi-asprino** [comment]
Ok.. now I got the same error.

**Comment by enridaga** [comment]
Some discussion (discouraging) here https://issues.apache.org/jira/browse/MNG-624 and here https://stackoverflow.com/questions/23493020/use-properties-value-to-set-maven-parent-version

The bottom line is that the parent pom is evoked *before* evaluating variables in the sub-module. So, the variable in the parent block of a sub-module is only evaluated if the sub-module is loaded *after* the reactor (that is also the parent in our project). 

However, the error refers to the dependency module sparql-anything-model's parent, not the parent of the current sub module (-cli). What is happening is that the built modules are placed in the local repository with a pom that cannot be adequately processed by maven (as the message reports).

The example on the Maven docs https://maven.apache.org/maven-ci-friendly.html warns about deploying artefacts with this setup as they will not be resolved by maven https://maven.apache.org/maven-ci-friendly.html#install-deploy unless one uses the Maven flatten plugin (which basically "flattens" all dependencies in the pom on deploy, thus avoiding leaving unresolved variables around).

**Comment by enridaga** [comment]
Fixed in [91400a0](https://github.com/SPARQL-Anything/sparql.anything/commit/91400a0c853b6674d65de0992873873f49dc79a1)
